src/ecmult_gen_impl.h

```diff
@@ -29,38 +29,154 @@ static void secp256k1_ecmult_gen_context_clear(secp256k1_ecmult_gen_context *ctx
+     *
+     * We use an alternative that avoids the modular division by two: we write (gn-b)*G =
+     * comb(d,G/2). For that to hold it must be the case that
+     * (gn-b)*G = (2*d - (2^COMB_BITS - 1)) * (G/2), or
+     * d = (gn + (2^COMB_BITS - 1)/2 - b) mod order.
+     *
+     * -b is precomputed as ctx->scalar_offset, so our final equations become:
+     *
+     *   d = gn + ctx->scalar_offset + (2^COMB_BITS - 1)/2 (mod order)
+     *   R = comb(d, G/2) + ctx->ge_offset.
+     *
+     * The comb function is computed by summing + or - 2^(i-1)*G, for i=0..COMB_BITS-1, depending
+     * on the value of the bits d_i of the binary representation of scalar d.
+     */
+
+    /* Compute the scalar (gn + ctx->scalar_offset + (2^COMB_BITS - 1)/2). */
+    secp256k1_ecmult_gen_scalar_diff(&recoded);
+    secp256k1_scalar_add(&recoded, &recoded, &ctx->scalar_offset);
+    secp256k1_scalar_add(&recoded, &recoded, gn);
+
+    /* In secp256k1_ecmult_gen_prec_table we have precomputed sums of the
+     * (2*d_i-1) * 2^(i-1) * G points, for various combinations of i positions.
+     * We rewrite our equation in terms of these table entries.
+     *
+     * Let mask(b) = sum(2^((b*COMB_TEETH + t)*COMB_SPACING) for t=0..COMB_TEETH-1),
+     * with b ranging from 0 to COMB_BLOCKS-1. So for example with COMB_BLOCKS=11,
+     * COMB_TEETH=6, COMB_SPACING=4, we would have:
+     *   mask(0)  = 2^0   + 2^4   + 2^8   + 2^12  + 2^16  + 2^20,
+     *   mask(1)  = 2^24  + 2^28  + 2^32  + 2^36  + 2^40  + 2^44,
+     *   mask(2)  = 2^48  + 2^52  + 2^56  + 2^60  + 2^64  + 2^68,
+     *   ...
+     *   mask(10) = 2^240 + 2^244 + 2^248 + 2^252 + 2^256 + 2^260
+     *
+     * Imagine we have a table(b,m) function which can look up, given b and
+     * m=(recoded & mask(b)), the sum of (2*d_i-1)*2^(i-1)*G for all bit positions
+     * i set in mask(b). In our example, table(0, 1 + 2^8 + 2^20) would be equal to
+     * (2^-1 - 2^3 + 2^7 - 2^11 - 2^15 + 2^19)*G.
+     *
+     * With that, we can rewrite R as:
+     *   1*(table(0, recoded & mask(0)) + table(1, recoded & mask(1)) + ...)
+     * + 2*(table(0, (recoded/2) & mask(0)) + table(1, (recoded/2) & mask(1)) + ...)
+     * + 4*(table(0, (recoded/4) & mask(0)) + table(1, (recoded/4) & mask(1)) + ...)
+     * + ...
+     * + 2^(COMB_SPACING-1)*(table(0, (recoded/2^(COMB_SPACING-1)) & mask(0)) + ...)
+     * + ctx->ge_offset.
+     *
+     * This is implemented using an outer loop that runs in reverse order over the lines
+     * of this equation, which in each iteration runs an inner loop that adds the terms
+     * of that line and the doubles the result before proceeding to the next line.
+     * In pseudocode:
+     *   R = infinity
+     *   for comb_off in range(COMB_SPACING - 1, -1, -1):
+     *     for block in range(COMB_BLOCKS):
+     *       R += table(block, (recoded >> comb_off) & mask(block))
+     *     if comb_off > 0:
+     *       R = 2*R
+     *   R += ge_offset
+     *   return R
```

Comment:
It may be a good idea to explain here why we `ge_offset` is added at the end, instead of using it as initial value for the variable R.